cpp/src/experimental/graph.cu

```diff
@@ -0,0 +1,552 @@
+      thrust::make_counting_iterator(vertex_t{0}),
+      detail::degree_from_offsets_t<vertex_t, edge_t>{offsets_.data()});
+
+    // optional expensive checks (part 2/2)
+
+    if (do_expensive_check) {
+      CUGRAPH_EXPECTS(thrust::is_sorted(rmm::exec_policy(default_stream)->on(default_stream),
+                                        degree_first,
+                                        degree_first + this->get_number_of_vertices(),
+                                        thrust::greater<edge_t>{}),
+                      "Invalid API parameter: sorted_by_degree is set to true, but degrees are not "
+                      "non-ascending.");
+    }
+
+    static_assert(detail::num_segments_per_vertex_partition == 3);
+    static_assert((detail::low_degree_threshold <= detail::mid_degree_threshold) &&
+                  (detail::mid_degree_threshold <= std::numeric_limits<edge_t>::max()));
+    rmm::device_uvector<edge_t> d_thresholds(detail::num_segments_per_vertex_partition - 1,
+                                             default_stream);
+    std::vector<edge_t> h_thresholds = {static_cast<edge_t>(detail::low_degree_threshold),
+                                        static_cast<edge_t>(detail::mid_degree_threshold)};
+    raft::update_device(
+      d_thresholds.data(), h_thresholds.data(), h_thresholds.size(), default_stream);
+
+    rmm::device_uvector<vertex_t> segment_offsets(detail::num_segments_per_vertex_partition + 1,
+                                                  default_stream);
+    segment_offsets.set_element_async(0, 0, default_stream);
+    segment_offsets.set_element_async(
+      detail::num_segments_per_vertex_partition, this->get_number_of_vertices(), default_stream);
+
+    thrust::upper_bound(rmm::exec_policy(default_stream)->on(default_stream),
+                        degree_first,
+                        degree_first + this->get_number_of_vertices(),
+                        d_thresholds.begin(),
+                        d_thresholds.end(),
+                        segment_offsets.begin() + 1);
+
+    segment_offsets_.resize(segment_offsets.size());
+    raft::update_host(
+      segment_offsets_.data(), segment_offsets.data(), segment_offsets.size(), default_stream);
+
+    CUDA_TRY(cudaStreamSynchronize(
+      default_stream));  // this is necessary as d_thresholds and segment_offsets will become
+                         // out-of-scpe once control flow exits this block and segment_offsets_ can
+                         // be used right after return.
+  }
+
+  // optional expensive checks (part 3/3)
+
+  if (do_expensive_check) {
+    // FIXME: check for symmetricity may better be implemetned with transpose().
+    if (is_symmetric) {}
+    // FIXME: check for duplicate edges may better be implemented after deciding whether to sort
+    // neighbor list or not.
+    if (!is_multigraph) {}
+  }
+}
+
+// explicit instantiation
+
+template class graph_t<uint32_t, uint32_t, float, true, true>;
```

Comment:
Related to 1, supporting integer needs more discussion. It may not matter much for SSSP but for PageRank, how should we handle division (a/b) if a is not an integer multiple of b? Should we really consider both integer & floating-point weights, or if users passes integer weights, should we just convert those to floating-point numbers?

2. Yeah... I guess we should use int32_t till those backward compatibility issues are resolved (IIRC, cuSparse is the main problem). In long run, better switch to uint32_t as this can cover nearly 2x more vertices.

Reply:
At first, if users pass integer weights we may just convert those to floating-point numbers. This does not really fly for SSSP and BC though as the largest `int` that can be stored properly in a `float` is about 16M. This not that big, especially when accumulating weights on a graph with billions of edges.

For 2), cusparse is one issue. More importantly, no cugraph structures and kernels have been updated or tested for unsigned.

Reply:
What do you mean by "stored properly"?

"IEEE 754 32-bit base-2 floating-point variable has a maximum value of (2 − 2^−23) × 2^127 ≈ 3.4028235 × 10^38"
https://en.wikipedia.org/wiki/Single-precision_floating-point_format
3.4028235 × 10^38 is way larger than 2^32 - 1.

floating point numbers cannot accurately store integers (except for few values like 0 and 1) and the distance between two closest numbers increases as numbers get large, but I guess this does not matter unless a graph has a really large diameter.

Reply:
Addressed 2, supporting integer wights needs more discussion.